File: Server/src/main/java/com/lfj/messenger/server/dao/MessageDAO.java
package com.lfj.messenger.server.dao;

import com.lfj.dev.annotations.ThreadSafe;
import com.lfj.messenger.dto.datatype.server.MessageDTO;
import com.lfj.messenger.dto.datatype.server.UserDTO;
import com.lfj.messenger.dto.request.GetMessageRequest;
import com.lfj.messenger.dto.request.MessageRequest;
import com.lfj.messenger.dto.request.GetMessageRequest.Direction;
import com.lfj.messenger.dto.types.MessageType;
import com.lfj.messenger.uuid7.UUIDv7;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.Timestamp;
import java.sql.ResultSet;
import java.sql.SQLException;

import java.time.Instant;
import java.util.Optional;
import java.util.Set;
import java.util.List;
import java.util.HashSet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.UUID;
import java.util.function.Supplier;

public class MessageDAO {
    private DataSource source;
    private Logger logger;
    private MessageDAO(){ this.logger = LoggerFactory.getLogger(MessageDAO.class); }

    private static final String WRITE_MESSAGE_TO_BD = "INSERT INTO messages_table(message_id, chat_id, sender_id, message_type, content, created_at) VALUES(?, ?, ?, ?, ?, ?)";
    private static final String READ_LATEST_MESSAGES = "SELECT message_id, chat_id, sender_id, message_type, content, created_at FROM messages_table WHERE chat_id = ? ORDER BY message_id DESC LIMIT ?";
    private static final String READ_OLDER_MESSAGES = "SELECT message_id, chat_id, sender_id, message_type, content, created_at FROM messages_table WHERE chat_id = ? AND message_id < ? ORDER BY message_id DESC LIMIT ?";
    private static final String GET_SENDER_ID_BY_CHAT_ID = "SELECT sender_id FROM messages_table WHERE chat_id = ? LIMIT ?";

    public MessageDAO(DataSource source){
        this();
        this.source = source;
    }
    @ThreadSafe
    public Supplier<Optional<MessageDTO>> writeAndSendMessageAsync(MessageRequest request){
        return () -> writeAndSendMessage(request);
    }
    @ThreadSafe
    public Supplier<Optional<List<MessageDTO>>> readNewerMessageAsync(GetMessageRequest request, Set<UserDTO> users){
        return () -> readNewerMessage(request, users);
    }
    @ThreadSafe
    public Supplier<Optional<List<MessageDTO>>> readOlderMessageAsync(GetMessageRequest request, Set<UserDTO> users){
        return () -> readOlderMessage(request, users);
    }
    @ThreadSafe
    public Supplier<Optional<List<MessageDTO>>> readMessageAsync(GetMessageRequest request, Set<UserDTO> users){
        if(request.direction() == Direction.NEWER) return () -> readNewerMessage(request, users);
        else return () -> readOlderMessage(request, users);
    }
    public Optional<MessageDTO> writeAndSendMessage(MessageRequest request){
        try(Connection connection = this.source.getConnection(); PreparedStatement stmt = connection.prepareStatement(WRITE_MESSAGE_TO_BD)) {
            UUID messageId = UUIDv7.next();
            Instant instant = Instant.now();
            stmt.setObject(1, messageId);
            stmt.setObject(2, request.getChatId());
            stmt.setObject(3, request.getSender().userId());
            stmt.setString(4, request.message().messageType().name());
            stmt.setString(5, request.message().content());
            stmt.setTimestamp(6, Timestamp.from(instant));
            stmt.executeUpdate();
            return Optional.ofNullable(message(messageId, request.getChatId(), request.getSender(), request.message().messageType(), request.message().content(), instant));
        } catch (SQLException e) {
            logger.error("{} - {}. State: {}", e.getMessage(), e.getErrorCode(), e.getSQLState());
            e.printStackTrace();
            return Optional.empty();
        }
    }
    public Optional<Set<UUID>> getSenderId(UUID chatId, int limit){
        Set<UUID> senderIds = new HashSet<>();
        try(Connection connection = this.source.getConnection(); PreparedStatement stmt = connection.prepareStatement(GET_SENDER_ID_BY_CHAT_ID)){
            stmt.setObject(1, chatId);
            stmt.setInt(2, limit);
            try(ResultSet rs = stmt.executeQuery()) {
                while (rs.next()){
                    UUID id = rs.getObject("sender_id", UUID.class);
                    if(!senderIds.contains(id)) senderIds.add(id);
                }
            }
        }catch (SQLException e){
            logger.error("Database error", e);
        }
        return Optional.of(senderIds);
    }

    public Optional<List<MessageDTO>> readNewerMessage(GetMessageRequest request, Set<UserDTO> users){
        if(request.direction() != GetMessageRequest.Direction.NEWER) return Optional.empty();

        try(Connection connection = this.source.getConnection(); PreparedStatement stmt = connection.prepareStatement(READ_LATEST_MESSAGES)){            List<MessageDTO> messages = new ArrayList<>();
            stmt.setObject(1, request.chatId());
            stmt.setInt(2, request.limit());
            try(ResultSet rs = stmt.executeQuery()){
                while(rs.next()){
                    UUID userId = rs.getObject("sender_id", UUID.class);
                    Optional<UserDTO> user = users.stream().filter(f -> f.userId().equals(userId)).findFirst();
                    if(user.isPresent()) messages.add(message(rs, user.get()));
                }
            }
            Collections.reverse(messages);
            return Optional.of(messages);
        }catch (SQLException e){
            logger.error("MessageDAO error", e);
        }
        return Optional.empty();
    }

    public Optional<List<MessageDTO>> readOlderMessage(GetMessageRequest request, Set<UserDTO> users){
        if(request.direction() != GetMessageRequest.Direction.OLDER) return Optional.empty();
        try(Connection connection = this.source.getConnection(); PreparedStatement stmt = connection.prepareStatement(READ_OLDER_MESSAGES)){
            stmt.setObject(1, request.chatId());
            stmt.setObject(2, request.cursorMessageId());
            stmt.setInt(3, request.limit());
            List<MessageDTO> messages = new ArrayList<>();
            try(ResultSet rs = stmt.executeQuery()){
                while (rs.next()){
                    UUID userId = rs.getObject("sender_id", UUID.class);
                    Optional<UserDTO> user = users.stream().filter(f -> f.userId().equals(userId)).findFirst();
                    if(user.isPresent()) messages.add(message(rs, user.get()));
                }
            }
            Collections.reverse(messages);
            return Optional.of(messages);
        }catch (SQLException e){
            logger.error("MessageDAO error", e);
        }
        return Optional.empty();
    }

    private MessageDTO message(ResultSet rs, UserDTO user) throws SQLException {
        return new MessageDTO(rs.getObject("message_id", UUID.class), rs.getObject("chat_id", UUID.class), user, MessageType.valueOf(rs.getString("message_type")), rs.getString("content"), rs.getTimestamp("created_at").toInstant());
    }

    private MessageDTO message(UUID messageId, UUID chatId, UserDTO user, MessageType messageType, String content, Instant instant){
        return new MessageDTO(messageId, chatId, user, messageType, content, instant);
    }
}

File: Server/src/main/kotlin/com/lfj/messfox/server/dao/MessageDao.kt
package com.lfj.messfox.server.dao

import com.lfj.messenger.uuid7.UUIDv7
import com.lfj.messfox.exceptions.MessFoxException
import com.lfj.messfox.protocol.type.MessageType
import com.lfj.messfox.server.ifPresentOrElseWithResult
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Timestamp
import java.time.Instant
import java.util.*
import javax.sql.DataSource
import kotlin.reflect.KClass

interface MessageDataTable{
    fun insertMessage(chatId: UUID, senderId: UUID, messageType: MessageType, content: String) : Result<MessageBillet>
    fun deleteMessage(messageId: UUID) : Boolean
    fun findMessageForId(messageId: UUID, chatId: UUID) : Result<MessageBillet>
    fun findMessagesForSenderId(senderId: UUID, chatId: UUID) : List<MessageBillet>
    fun getLastMessage(chatId: UUID) : Result<MessageBillet>

    interface CompanionsSQLRequest{
        val INSERT_MESSAGE: String
        val DELETE_MESSAGE: String
        val FIND_MESSAGE_FOR_ID: String
        val FIND_MESSAGES_FOR_SENDER_ID: String
        val GET_LAST_MESSAGE: String
        val GET_LATEST_MESSAGES: String
        val GET_ORDER_MESSAGES: String
    }

    fun mapToMessageBillet(rs: ResultSet) : MessageBillet{
        return MessageBillet(rs.getObject("message_id", UUID::class.java), rs.getObject("chat_id", UUID::class.java), rs.getObject("sender_id", UUID::class.java), MessageType.valueOf(rs.getString("message_type")), rs.getString("content"), rs.getTimestamp("created_at").toInstant())
    }

    open class AbstractMessageDAO <T : MessageDataTable> (open val dataSource: DataSource, val sqlRequest: CompanionsSQLRequest, clazz: KClass<T>) : MessageDataTable{
        private val logger: Logger = LoggerFactory.getLogger(clazz.java)
        override fun insertMessage(chatId: UUID, senderId: UUID, messageType: MessageType, content: String): Result<MessageBillet> {
            return try {
                dataSource.connection.use { conn ->
                    conn.prepareStatement(sqlRequest.INSERT_MESSAGE).use { stmt ->
                        val messageId: UUID = UUIDv7.next()
                        UUIDv7.getTimestamp(messageId).ifPresentOrElseWithResult(action = {
                            stmt.setObject(1, messageId)
                            stmt.setObject(2, chatId)
                            stmt.setObject(3, senderId)
                            stmt.setString(4, messageType.name)
                            stmt.setString(5, content)
                            stmt.setTimestamp(6, Timestamp.from(it))
                            stmt.executeUpdate()
                            Result.success(MessageBillet(messageId, chatId, senderId, messageType, content, it))
                        }, emptyAction = { Result.failure(IllegalStateException("")) })
                    }
                }
            }catch (e: SQLException){
                logger.error("MessageDao error", e)
                Result.failure(e)
            }
        }

        override fun deleteMessage(messageId: UUID) : Boolean {
            try { // Добавить проверку идентификатора отправителя
                dataSource.connection.use { conn ->
                    conn.prepareStatement(sqlRequest.DELETE_MESSAGE).use { stmt ->
                        stmt.setObject(1, messageId)
                        stmt.executeUpdate()
                        return true
                    }
                }
            }catch (e: SQLException){
                logger.error("MessageDAO Error", e)
                return false
            }
        }

        override fun findMessageForId(messageId: UUID, chatId: UUID): Result<MessageBillet> {
            return try {
                dataSource.connection.use { conn ->
                    conn.prepareStatement(sqlRequest.FIND_MESSAGE_FOR_ID).use { stmt ->
                        stmt.setObject(1, chatId)
                        stmt.setObject(2, messageId)
                        stmt.executeQuery().use { rs ->
                            if (rs.next()) {
                                logger.debug("Message find by id SUCCESS")
                                Result.success(mapToMessageBillet(rs))
                            } else {
                                logger.debug("Message find by id FAILED")
                                Result.failure(IllegalStateException(""))
                            }
                        }
                    }
                }
            }catch (e: SQLException){
                logger.error("MessageDAO error", e)
                Result.failure(e)
            }
        }

        override fun findMessagesForSenderId(senderId: UUID, chatId: UUID): List<MessageBillet> {
            return try{
                val mutableList: MutableList<MessageBillet> = mutableListOf()
                dataSource.connection.use { conn ->
                    conn.prepareStatement(sqlRequest.FIND_MESSAGES_FOR_SENDER_ID).use { stmt ->
                        stmt.setObject(1, chatId)
                        stmt.setObject(2, senderId)
                        stmt.executeQuery().use{ rs ->
                            while (rs.next()){
                                mutableList.add(mapToMessageBillet(rs))
                            }
                        }
                        mutableList.toList()
                    }
                }
            }catch (e: SQLException){
                logger.error("MessageDAO error", e)
                listOf()
            }
        }
        override fun getLastMessage(chatId: UUID) : Result<MessageBillet> {
            return try{
                dataSource.connection.use { conn ->
                    conn.prepareStatement(sqlRequest.GET_LAST_MESSAGE).use { stmt ->
                        stmt.setObject(1, chatId)
                        stmt.executeQuery().use { rs ->
                            if(rs.next()) Result.success(mapToMessageBillet(rs))
                            else Result.failure(MessFoxException("Message not found!"))
                        }
                    }
                }
            }catch (e: SQLException){
                logger.error("MessageDAO error", e)
                Result.failure(MessFoxException("Message not found!", e))
            }
        }
    }

    class H2MessageDAO(override val dataSource: DataSource) : AbstractMessageDAO<H2MessageDAO>(dataSource, h2SQLRequest, H2MessageDAO::class){
        private companion object h2SQLRequest : CompanionsSQLRequest{
            override val INSERT_MESSAGE = """
                INSERT INTO messages_table(message_id, chat_id, sender_id, message_type, content, created_at)
                VALUES (?, ?, ?, ?, ?, ?)
            """.trimIndent()
            override val DELETE_MESSAGE = """
                DELETE FROM messages_table
                WHERE message_id = ? AND user_id OR ? 
            """.trimIndent() // ???
            override val FIND_MESSAGE_FOR_ID = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ? AND message_id = ?
            """.trimIndent()
            override val FIND_MESSAGES_FOR_SENDER_ID = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ? AND sender_id = ?
                LIMIT 19
            """.trimIndent()
            override val GET_LAST_MESSAGE: String = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ?
                ORDER BY message_id DESC
                LIMIT 1 
            """.trimIndent()
            override val GET_LATEST_MESSAGES: String = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ?
                ORDER BY message_id DESC
                LIMIT 50 
            """.trimIndent()
            override val GET_ORDER_MESSAGES: String = """
                SELECT message_id, chat_id. sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ? AND message_id < ?
                LIMIT 50 
            """.trimIndent()
        }
    }
    class PostgresMessageDAO(override val dataSource: DataSource) : AbstractMessageDAO<PostgresMessageDAO>(dataSource, postgreSQLRequest, PostgresMessageDAO::class){
        private companion object postgreSQLRequest : CompanionsSQLRequest{
            override val INSERT_MESSAGE = """
                INSERT INTO messages_table(message_id, chat_id, sender_id, message_type, content, created_at)
                VALUES (?, ?, ?, ?, ?, ?)
                ON CONFLICT(message_id) DO NOTHING
            """.trimIndent()
            override val DELETE_MESSAGE = """
                DELETE FROM messages_table
                WHERE message_id = ?
            """.trimIndent()
            override val FIND_MESSAGE_FOR_ID = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ? AND message_id = ?
            """.trimIndent()
            override val FIND_MESSAGES_FOR_SENDER_ID = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ? AND sender_id = ?
                LIMIT 19
            """.trimIndent()
            override val GET_LAST_MESSAGE: String = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ?
                ORDER BY message_id DESC
                LIMIT 1 
            """.trimIndent()
            override val GET_LATEST_MESSAGES: String = """
                SELECT message_id, chat_id, sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ?
                ORDER BY message_id DESC
                LIMIT 50 
            """.trimIndent()
            override val GET_ORDER_MESSAGES: String = """
                SELECT message_id, chat_id. sender_id, message_type, content, created_at
                FROM messages_table
                WHERE chat_id = ? AND message_id < ?
                LIMIT 50 
            """.trimIndent()
        }
    }
    data class MessageBillet(val messageId: UUID, val chatId: UUID, val senderId: UUID, val messageType: MessageType, val content: String, val createdAt: Instant)
}